Enhanced Laurent polynomial divisibility and ideal membership using cover rings

Hapsa21 opened the pull request:
### Optimized Laurent polynomial membership checks

#### Summary
This PR fixes divisibility and ideal membership for Laurent polynomials over rings with zero-divisors. It preserves fast paths for integral domains and uses a cover ring isomorphism for other cases.

#### Method B Implementation
* **Approach**: Uses a single-variable isomorphism $R[x_1, \dots, x_n, T] / (T \cdot \prod x_i - 1)$.
* **Performance**: Faster than adding multiple variables (~370 µs vs ~580 µs for $N=4$).
* **Logic**: Localization at $\prod x_i$ makes each $x_i$ invertible, ensuring the map is bijective.



#### Bug Fixes
* **AttributeError**: Switched to `monomial_reduction()` instead of the non-existent `polynomial_part`.
* **TypeError**: Used `polynomial_ideal()` to fix a lazy-initialization bug.

#### Verification
* All tests in `src/sage/rings/polynomial/laurent_polynomial*` pass locally.

fixes #41318

Hapsa21 committed:
Fix #41318: Improve Laurent polynomial .divides() and ideal membership using cover rings

Hapsa21 committed:
Fix linting issues in laurent_polynomial_ring.py

enriqueartal commented:
As you said, Laurent polynomials have distinct approaches, and part of the code is quite old. I made some changes at some point without changing the structure since I feared to introduce errors elsewhere.

I did not check in detail your changes. Are previous algorithms still available, eventually with keywords, for cases where the base ring is a domain (or a field)? 
It would be nice to compare times when various approaches are available. Another point, you can also use the ideal generated by $Tx_1\dots x_n - 1$ adding only a variable $T$? Is it computationally worse?

user202729 commented:
To save readers' time, only include important detail in the pull request description. (Overly verbose pull request description is usually a symptom of LLM over-reliance.)

Hapsa21 commented:
> As you said, Laurent polynomials have distinct approaches, and part of the code is quite old. I made some changes at some point without changing the structure since I feared to introduce errors elsewhere.
> 
> I did not check in detail your changes. Are previous algorithms still available, eventually with keywords, for cases where the base ring is a domain (or a field)? It would be nice to compare times when various approaches are available. Another point, you can also use the ideal generated by T x 1 … x n − 1 adding only a variable T ? Is it computationally worse?

I updated the code to a hybrid approach. For integral domains, it still uses the old fast algorithms so there's no speed drop. I found using $x_iy_i - 1$ more direct for handling zero-divisors than the single $T$ variable approach. I also fixed that multivariate AttributeError and all tests pass now.

Hapsa21 committed:
Merge branch 'develop' into fix-41318-laurent-divides

user202729 commented:
> you can also use the ideal generated by T x 1 … x n − 1 adding only a variable T ? Is it computationally worse?

Actually how do you prove that the map is bijective? For N variables it isn't very obvious right? (haven't thought too carefully about it)

Computationally, it might be better that way (complexity is probably exponential in number of variables, so less variables is better even at the cost of polynomially larger degree), but benchmarking needed.

Hapsa21 commented:
> > you can also use the ideal generated by T x 1 … x n − 1 adding only a variable T ? Is it computationally worse?
> 
> Actually how do you prove that the map is bijective? For N variables it isn't very obvious right? (haven't thought too carefully about it)
> 
> Computationally, it might be better that way (complexity is probably exponential in number of variables, so less variables is better even at the cost of polynomially larger degree), but benchmarking needed.

<img width="519" height="236" alt="image" src="https://github.com/user-attachments/assets/9500a88c-630d-4aae-9899-70dc8a9f1f31" />


I have updated the PR to use the single-variable $T$ approach as you suggested. As you suspected, Method B is more efficient even for small $N$. I ran multiple benchmarks, the single-variable approach consistently outperforms the multi-variable one, as shown in the attached screenshot. 

Regarding bijectivity: The map is a standard isomorphism for localizing a ring at the product of its variables, $f = \prod x_i$. Making the product invertible in the cover ring $R[x_1, \dots, x_n, T]/(Tf - 1)$ makes each $x_i$ invertible, which correctly models the Laurent ring. The inverse map is well-defined by $T \mapsto (x_1 \dots x_n)^{-1}$.

user202729 commented:
- Don't delete tests.
- Don't add exception handling.

Hapsa21 commented:
> * Don't delete tests.
> * Don't add exception handling.

done

user202729 commented:
- need more tests to test the newly added code branches.

    (Don't use LLM to write tests. Make sure you understand what the **problem** is and that you can compute the expected output by hand. If you don't understand the problem, say you don't know.)

- the `hasattr(f, 'monomial_reduction')` is suspicious. Just cast `f` to `self.ring()` like the old code.
- make sure the documentation stands by itself. `Method B` doesn't make sense by itself.
- format docstrings with reStructuredText.

Hapsa21 commented:
> * need more tests to test the newly added code branches.
>   (Don't use LLM to write tests. Make sure you understand what the **problem** is and that you can compute the expected output by hand. If you don't understand the problem, say you don't know.)
> * the `hasattr(f, 'monomial_reduction')` is suspicious. Just cast `f` to `self.ring()` like the old code.
> * make sure the documentation stands by itself. `Method B` doesn't make sense by itself.
> * format docstrings with reStructuredText.

 @user202729  i understood and have added these tests kindly have a look

Hapsa21 committed:
Merge branch 'develop' into fix-41318-laurent-divides

Hapsa21 committed:
Merge branch 'develop' into fix-41318-laurent-divides

user202729 commented on src/sage/rings/polynomial/laurent_polynomial.pyx:
```diff
@@ -2239,16 +2254,34 @@ cdef class LaurentPolynomial_univariate(LaurentPolynomial):
         """
-        if self.base_ring().is_integral_domain() is True:
-            p = self.polynomial_construction()[0]
-            q = other.polynomial_construction()[0]
+        if self.is_zero():
+            return other.is_zero()
```
what happens if you delete these 2 lines?

Hapsa21 replied:
On deletion the code would proceed to the cover ring ideal check then it might return a correct result mathematically but i kept these lines to maintain efficiency and explicitly follow the standard sagemath handling for zero-divisibility cases, avoiding unnecessary ring constructions

user202729 commented on src/sage/rings/polynomial/laurent_polynomial_ring.py:
```diff
@@ -574,15 +583,25 @@ def __init__(self, R):
+        Return the polynomial cover ring with a single variable T.
+        """
+        base_names = self.variable_names()
+        new_names = base_names + ('T',)
```
Nothing in the current code guarantees that `'T' not in base_names`

Hapsa21 replied:
> Nothing in the current code guarantees that `'T' not in base_names`

i've updated `_poly_cover_ring` to dynamically find a unique name by checking against `base_names` and incrementing (e.g. T, T0, T1.....) until there is no conflict I also added a test case to verify this behavior.

Hapsa21 committed:
Merge branch 'develop' into fix-41318-laurent-divides

Hapsa21 committed:
Merge branch 'develop' into fix-41318-laurent-divides

Hapsa21 committed:
Merge branch 'develop' into fix-41318-laurent-divides
